### DISCUSSION TITLE
solutions unconverged for porous structures

### DISCUSSION THREAD

**Discussion by chenyuanming223** [idaholab/moose#24586] [Q&A Modules: Solid mechanics] [answered]
Dear MOOSE users:

I want to simulate the uniaxial comp with porous lattice structures using moose. The material model is elastic-plastic model with a small hardening for better convergence. but during simulation, the time-stepping is always reducing the minimal, then stopped for both NEWTON and PJFNK. The input files are attached, could you please help me look at this and give me any suggestions?
Thanks!
[input_files_zc.zip](https://github.com/idaholab/moose/files/11645234/input_files_zc.zip)

ZC

**Comment by GiudGiud** [comment]
Hello

Please see our page for troubleshooting failing solves:

https://mooseframework.inl.gov/application_usage/failed_solves.html

Guillaume

**Reply by chenyuanming223** [reply]
Hi Guillaume,
thanks, for a simple structure generated by moose mesh generator, this input script is good.

**Reply by GiudGiud** [reply]
Interesting. 
Do you mind hosting these files, especially the mesh, on a githup repo? Zip isn't a great format for us

**Reply by chenyuanming223** [reply]
Hi,
Please see this: https://github.com/chenyuanming223/test_moose.git

**Reply by hugary1995** [reply]
Don't use `compute_scaling_once=false` it's a bad option. Use preset=false on BCs for plasticity models. Use Newton not PJFNK since the Jacobian is exact.

**Reply by hugary1995** [reply]
Linear tet is not a good choice for plasticity either.

**Reply by hugary1995** [reply]
You also need to fiddle with the solver. The default solver is not going to work the best.

**Reply by chenyuanming223** [reply]
> You also need to fiddle with the solver. The default solver is not going to work the best.

Hi Gary,

You're right, the nolinear convergence residual is too tight I think, compared with Ansys mechanical (the default relative residual for Force is 0.5%). In moose, we have nl_abs_tol and nl_rel_tol, the solver in this simulation checks according to nl_abs_tol?

**Reply by chenyuanming223** [reply]
> You also need to fiddle with the solver. The default solver is not going to work the best.

with the hex8, still the error: 
We caught a libMesh error in ThreadedElementLoopBase:ERROR: negative Jacobian -2.09741e-06 at point (x,y,z)=( 4.73366,  4.93641, 0.548187) in element 54113
[7] ../src/fe/fe_map.C, line 1173, compiled Apr 20 2023 at 15:32:26
Nonlinear solve did not converge due to DIVERGED_LINE_SEARCH iterations 0
then timestepdroped

**Reply by GiudGiud** [reply]
>  the solver in this simulation checks according to nl_abs_tol?

It checks based on both, and the nl_abs_tol default is 1e-50, so it s usually the nl_rel_tol that gets triggered

Is the HEX8 mesh working on the first time step?

Guillaume

**Reply by hugary1995** [reply]
If you can upload your hex mesh to that repo, I can quickly make the changes for you. I really should add a troubleshoot page for tensor mechanics...

**Reply by chenyuanming223** [reply]
> If you can upload your hex mesh to that repo, I can quickly make the changes for you. I really should add a troubleshoot page for tensor mechanics...

Hi Gary,

Thanks, the hex mesh was uploaded.

**Reply by hugary1995** [reply]
Here,

```python
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
[]

[Mesh]
  type = FileMesh
  file = BCC10_441_hex.msh
  use_displaced_mesh = false
[]

[Modules]
  [TensorMechanics]
    [Master]
      [all]
        volumetric_locking_correction = true
        formulation = TOTAL
        new_system = true
        add_variables = true
        strain = FINITE
      []
    []
  []
[]

[BCs]
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bot'
    value = 0
    preset = false
  []
  [top_y]
    type = FunctionDirichletBC
    boundary = 'top'
    function = -.02*t
    variable = disp_y
    preset = false
  []
  [left_x]
    type = DirichletBC
    value = 0
    variable = disp_x
    boundary = 'left'
  []
  [back_z]
    type = DirichletBC
    value = 0
    variable = disp_z
    boundary = 'back'
  []
[]

[Materials]
  [stress]
    type = ComputeMultipleInelasticStress
    inelastic_models = 'isoplas'
  []
  [isoplas]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 880
    hardening_constant = 1000
    max_inelastic_increment = 0.1
  []
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1.93e5
    poissons_ratio = 0.3
  []
  [compute_stress]
    type = ComputeLagrangianWrappedStress
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -pc_hypre_type -ksp_gmres_restart -pc_hypre_boomeramg_strong_threshold -pc_hypre_boomeramg_interp_type -pc_hypre_boomeramg_coarsen_type -pc_hypre_boomeramg_agg_nl -pc_hypre_boomeramg_agg_num_paths -pc_hypre_boomeramg_truncfactor'
  petsc_options_value = 'hypre boomeramg 301 0.7 ext+i PMIS 4 2 0.4'
  line_search = none
  end_time = 100
  dtmin = 1e-10
  dtmax = 2
  automatic_scaling = true
  nl_max_its = 15
  nl_abs_tol = 1e-8
  nl_rel_tol = 1e-6
  l_max_its = 300
  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 6
    iteration_window = 1
    dt = 1e-2
    cutback_factor = .75
    growth_factor = 1.5
  []
[]

[Outputs]
  exodus = true
  csv = true
[]
```

I'll leave it to you to diff your original input deck against this modified one. Feel free to ask if you have further questions related to the changes.

**Reply by chenyuanming223** [reply]
Hi Gary,

Thanks, I test this code, still the timestep will drop to 1e-4, makes the simulation very slow. Besides, I use mpiexec with 40 cores on the campus cluster, but the speed didn't improve much

**Reply by hugary1995** [reply]
Can you paste the convergence log here? IIRC I tested my modified input file and it worked well. I didn't let it run till the end though, so it may be failing to converge at a later step.

**Reply by hugary1995** [reply]
Just save the console output in a text file and upload it here.

**Reply by chenyuanming223** [reply]
Hi Gary,

see this https://github.com/chenyuanming223/test_moose.git

**Reply by hugary1995** [reply, marked as answer]
Uhh that one probably doesn't have exact Jacobian. Use a different plasticity model then...
```python
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
[]

[Mesh]
  type = FileMesh
  file = BCC10_441_hex.msh
  use_displaced_mesh = false
[]

[Modules]
  [TensorMechanics]
    [Master]
      [all]
        volumetric_locking_correction = true
        formulation = TOTAL
        new_system = true
        add_variables = true
        strain = FINITE
      []
    []
  []
[]

[BCs]
  [bottom_y]
    type = DirichletBC
    variable = disp_y
    boundary = 'bot'
    value = 0
  []
  [top_y]
    type = FunctionDirichletBC
    boundary = 'top'
    function = -.02*t
    variable = disp_y
    preset = false
  []
  [left_x]
    type = DirichletBC
    value = 0
    variable = disp_x
    boundary = 'left'
  []
  [back_z]
    type = DirichletBC
    value = 0
    variable = disp_z
    boundary = 'back'
  []
[]

[Materials]
  # [isoplas]
  #   type = IsotropicPlasticityStressUpdate
  #   yield_stress = 880
  #   hardening_constant = 1000
  #   max_inelastic_increment = 0.1
  #   automatic_differentiation_return_mapping = true
  # []
  # [compute_stress]
  #   type = ComputeLagrangianWrappedStress
  # []  
  # [stress]
  #   type = ComputeMultipleInelasticStress
  #   inelastic_models = 'isoplas'
  # []
  [elasticity]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1.93e5
    poissons_ratio = 0.3
  []
  [flow_stress]
    type = DerivativeParsedMaterial
    property_name = flow_stress
    expression = '800+1000*effective_plastic_strain'
    material_property_names = 'effective_plastic_strain'
    additional_derivative_symbols = 'effective_plastic_strain'
    derivative_order = 2
    compute = false
  []
  [compute_stress]
    type = ComputeSimoHughesJ2PlasticityStress
    flow_stress_material = flow_stress
    output_properties = 'effective_plastic_strain'
    outputs = 'exodus'
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options = '-ksp_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -ksp_type'
  petsc_options_value = 'lu       superlu_dist                  gmres'
  reuse_preconditioner = true
  reuse_preconditioner_max_linear_its = 50
  line_search = none
  end_time = 100
  dtmin = 1e-10
  dtmax = 2
  automatic_scaling = true
  nl_max_its = 15
  nl_abs_tol = 1e-8
  nl_rel_tol = 1e-6
  l_max_its = 300
  [TimeStepper]
    type = IterationAdaptiveDT
    optimal_iterations = 6
    iteration_window = 1
    dt = 1e-2
    cutback_factor = .75
    growth_factor = 1.5
  []
  [Predictor]
    type = SimplePredictor
    scale = 1
  []
[]

[Outputs]
  print_linear_residuals = false
  exodus = true
  csv = true
[]
```

I just tested -- This one can run to completion.

![image](https://github.com/idaholab/moose/assets/31555204/254d26ba-9b5e-4def-9d0b-f35d072119a2)

**Reply by hugary1995** [reply]
oops I made a typo in the yield stress: should be 880 instead of 800. But anyways you get the idea.

**Reply by chenyuanming223** [reply]
Hi Gary,

Thanks, very appreciate your kindness, I'll test this code
